File: binding/src/libqsavefile.cpp
#include <QFileDevice>
#include <QObject>
#include <QSaveFile>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qsavefile.h>
#include "libqsavefile.h"
#include "libqsavefile.hxx"

QSaveFile* QSaveFile_new(const libqt_string name) {
	 return new VirtualQSaveFile(QString::fromUtf8(name.data, name.len));
}

QSaveFile* QSaveFile_new2() {
	 return new VirtualQSaveFile();
}

QSaveFile* QSaveFile_new3(const libqt_string name, QObject* parent) {
	 return new VirtualQSaveFile(QString::fromUtf8(name.data, name.len), parent);
}

QSaveFile* QSaveFile_new4(QObject* parent) {
	 return new VirtualQSaveFile(parent);
}

libqt_string QSaveFile_Tr(const char* s) {
	QString _ret = QSaveFile::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSaveFile_FileName(const QSaveFile* self) {
	QString _ret = self->fileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSaveFile_SetFileName(QSaveFile* self, const libqt_string name) {
	self->setFileName(QString::fromUtf8(name.data, name.len));
}

bool QSaveFile_Open(QSaveFile* self, int flags) {
	return self->open(static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(flags));
}

bool QSaveFile_Commit(QSaveFile* self) {
	return self->commit();
}

void QSaveFile_CancelWriting(QSaveFile* self) {
	self->cancelWriting();
}

void QSaveFile_SetDirectWriteFallback(QSaveFile* self, bool enabled) {
	self->setDirectWriteFallback(enabled);
}

bool QSaveFile_DirectWriteFallback(const QSaveFile* self) {
	return self->directWriteFallback();
}

libqt_string QSaveFile_Tr2(const char* s, const char* c) {
	QString _ret = QSaveFile::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSaveFile_Tr3(const char* s, const char* c, int n) {
	QString _ret = QSaveFile::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
libqt_string QSaveFile_QBaseFileName(const QSaveFile* self) {
	auto* vqsavefile = dynamic_cast<const VirtualQSaveFile*>(self);
	if (vqsavefile && vqsavefile->isVirtualQSaveFile) {
vqsavefile->setQSaveFile_FileName_IsBase(true);
	QString _ret = vqsavefile->fileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QSaveFile_OnFileName(const QSaveFile* self, intptr_t slot) {
	auto* vqsavefile = dynamic_cast<const VirtualQSaveFile*>(self);
	if (vqsavefile && vqsavefile->isVirtualQSaveFile) {
vqsavefile->setQSaveFile_FileName_Callback(reinterpret_cast<VirtualQSaveFile::QSaveFile_FileName_Callback>(slot));
}
}

// Base class handler implementation
bool QSaveFile_QBaseOpen(QSaveFile* self, int flags) {
	auto* vqsavefile = dynamic_cast<VirtualQSaveFile*>(self);
	if (vqsavefile && vqsavefile->isVirtualQSaveFile) {
vqsavefile->setQSaveFile_Open_IsBase(true);
	return vqsavefile->open(static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(flags));
}
}

// Auxiliary method to allow providing re-implementation
void QSaveFile_OnOpen(QSaveFile* self, intptr_t slot) {
	auto* vqsavefile = dynamic_cast<VirtualQSaveFile*>(self);
	if (vqsavefile && vqsavefile->isVirtualQSaveFile) {
vqsavefile->setQSaveFile_Open_Callback(reinterpret_cast<VirtualQSaveFile::QSaveFile_Open_Callback>(slot));
}
}

// Derived class handler implementation
long long QSaveFile_WriteData(QSaveFile* self, const char* data, long long lenVal) {
	auto* vqsavefile = dynamic_cast<VirtualQSaveFile*>(self);
	if (vqsavefile && vqsavefile->isVirtualQSaveFile) {
	return vqsavefile->writeData(data, lenVal);
	} else {
	return self->QSaveFile::writeData(data, lenVal);
}
}

// Base class handler implementation
long long QSaveFile_QBaseWriteData(QSaveFile* self, const char* data, long long lenVal) {
	auto* vqsavefile = dynamic_cast<VirtualQSaveFile*>(self);
	if (vqsavefile && vqsavefile->isVirtualQSaveFile) {
vqsavefile->setQSaveFile_WriteData_IsBase(true);
	return vqsavefile->writeData(data, lenVal);
}
}

// Auxiliary method to allow providing re-implementation
void QSaveFile_OnWriteData(QSaveFile* self, intptr_t slot) {
	auto* vqsavefile = dynamic_cast<VirtualQSaveFile*>(self);
	if (vqsavefile && vqsavefile->isVirtualQSaveFile) {
vqsavefile->setQSaveFile_WriteData_Callback(reinterpret_cast<VirtualQSaveFile::QSaveFile_WriteData_Callback>(slot));
}
}

void QSaveFile_Delete(QSaveFile* self) {
    delete self;
}

File: binding/include/libqsavefile.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsavefile.html
pub const qsavefile = struct {

    /// New constructs a new QSaveFile object.
    pub fn New(name: []const u8) QtC.QSaveFile {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QSaveFile_new(name_str);
    }


    /// New2 constructs a new QSaveFile object.
    pub fn New2() QtC.QSaveFile {
        return qtc.QSaveFile_new2();
    }


    /// New3 constructs a new QSaveFile object.
    pub fn New3(name: []const u8, parent: ?*anyopaque) QtC.QSaveFile {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QSaveFile_new3(name_str, @ptrCast(parent));
    }


    /// New4 constructs a new QSaveFile object.
    pub fn New4(parent: ?*anyopaque) QtC.QSaveFile {
        return qtc.QSaveFile_new4(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSaveFile_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsavefile.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSaveFile_FileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsavefile.FileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnFileName(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.QSaveFile_OnFileName(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseFileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSaveFile_QBaseFileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsavefile.FileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetFileName(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
qtc.QSaveFile_SetFileName(@ptrCast(self), name_str);
    }

    pub fn Open(self: ?*anyopaque, flags: i32) bool {
        return qtc.QSaveFile_Open(@ptrCast(self), @intCast(flags));
    }

    /// Allows for overriding the related default method
    pub fn OnOpen(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) bool) void {
        qtc.QSaveFile_OnOpen(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseOpen(self: ?*anyopaque, flags: i32) bool {
        return qtc.QSaveFile_QBaseOpen(@ptrCast(self), @intCast(flags));
    }

    pub fn Commit(self: ?*anyopaque, ) bool {
        return qtc.QSaveFile_Commit(@ptrCast(self));
    }

    pub fn CancelWriting(self: ?*anyopaque, ) void {
        qtc.QSaveFile_CancelWriting(@ptrCast(self));
    }

    pub fn SetDirectWriteFallback(self: ?*anyopaque, enabled: bool) void {
        qtc.QSaveFile_SetDirectWriteFallback(@ptrCast(self), enabled);
    }

    pub fn DirectWriteFallback(self: ?*anyopaque, ) bool {
        return qtc.QSaveFile_DirectWriteFallback(@ptrCast(self));
    }

    pub fn WriteData(self: ?*anyopaque, data: []const u8, lenVal: i64) i64 {
        return qtc.QSaveFile_WriteData(@ptrCast(self), @ptrCast(data), lenVal);
    }

    /// Allows for overriding the related default method
    pub fn OnWriteData(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8, i64) callconv(.c) i64) void {
        qtc.QSaveFile_OnWriteData(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseWriteData(self: ?*anyopaque, data: []const u8, lenVal: i64) i64 {
        return qtc.QSaveFile_QBaseWriteData(@ptrCast(self), @ptrCast(data), lenVal);
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSaveFile_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsavefile.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSaveFile_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsavefile.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSaveFile_Delete(@ptrCast(self));
    }
};
